lib/core/src/Cardano/Wallet/DB/Store/Wallets/Model.hs

```diff
@@ -88,49 +91,88 @@ import qualified Data.Set as Set
 
 data DeltaTxWalletsHistory
     = ExpandTxWalletsHistory W.WalletId [(WT.Tx, WT.TxMeta)]
-    | ChangeTxMetaWalletsHistory W.WalletId ManipulateTxMetaHistory
+    | ChangeTxMetaWalletsHistory W.WalletId DeltaWalletsMetaWithSubmissions
     | GarbageCollectTxWalletsHistory
     | RemoveWallet W.WalletId
     deriving ( Show, Eq )
 
 instance Buildable DeltaTxWalletsHistory where
     build = build . show
 
-type TxWalletsHistory = (TxHistory, Map W.WalletId TxMetaHistory)
+data DeltaWalletsMetaWithSubmissions
+    = ChangeMeta DeltaTxMetaHistory
+    | ChangeSubmissions DeltaTxLocalSubmission
+    deriving ( Show, Eq )
+
+type MetasAndSubmissionsHistory = (TxMetaHistory, TxLocalSubmissionHistory)
+
+constraintSubmissions
+    :: MetasAndSubmissionsHistory -> MetasAndSubmissionsHistory
+constraintSubmissions (metas,submissions) =
+    ( metas
+    , over #relations (\m -> Map.restrictKeys m
+                       $ Map.keysSet (metas ^. #relations)) submissions)
+
+instance Delta DeltaWalletsMetaWithSubmissions where
+    type Base DeltaWalletsMetaWithSubmissions = MetasAndSubmissionsHistory
+    apply (ChangeMeta cm) (metas,submissions) =
+        constraintSubmissions (apply cm metas, submissions)
+    apply (ChangeSubmissions cs) (metas,submissions) =
+        constraintSubmissions (metas, apply cs submissions)
+
+embedConstrainedSubmissions :: Embedding'
+        DeltaWalletsMetaWithSubmissions
+        (DeltaTxMetaHistory, DeltaTxLocalSubmission)
+embedConstrainedSubmissions =
+    Embedding'
+    { load = Right
+    , write = id
+    , update = \_ _ -> \case
+          ChangeMeta dtmh -> (dtmh, TxSubmissions.Expand mempty)
+          ChangeSubmissions dtls -> (TxMetaStore.Expand mempty, dtls)
```

Comment:
Alas, this does not seem correct to me: It's not enough to apply the delta to one element of the pair, we also have to compute a delta that corresponds to the effect of `constraintSubmissions`. In other words, updating with `ChangeMeta` may indeed involve an additional ` TxSubmissions.Prune`.

From my comment at `removePendingOrExpiredTx`, I'm beginning to think that the implementation becomes simpler if we do *not* use a constraint, i.e. if we drop the `constraintSubmissions` entirely.
— Alternatively, we can make `embedConstrainedSubmissions` with constraints the "source of truth" and implement `apply` in terms of that.